### PR TITLE
autodoc: include entire code on line in source

### LINKED ISSUE #19293: better handle indentation when rendering source code
Extracted from #19249.

[Example](https://ziglang.org/documentation/master/std/#std.array_hash_map.ArrayHashMapUnmanaged.count)

```zig
pub fn count(self: Self) usize {
            return self.entries.len;
        }
```

Instead it should render like this:

```zig
pub fn count(self: Self) usize {
    return self.entries.len;
}
```

This happens because it is rendering only the function declaration directly from the file, which in fact is indented two levels. The rendering function needs to detect this and chop off the indentation levels.

#### Case 1

Input:
```zig
    fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 2

Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 3

Input: 
```zig
    fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:

```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

#### Case 4

Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

Note that those last two cases are planned to become syntax errors:
* #35

### PR DESCRIPTION
**Pull request by sno2**
Closes #19293

I took the easy route and just add the entire line's contents. This fixes the noted issue:

image

> However, care should be taken to make sure it does not introduce bugs if there is non-whitespace on that same line.

I am unsure if I introduced what you were warning against. Do you want

```zig
const a = 1; fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

to render as itself, or

```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

or

```zig
             fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

This implementation assumes you want the first option.

### CODE REVIEW

File: lib/docs/wasm/main.zig [hunk 1]
```diff
@@ -540,7 +542,9 @@ export fn decl_doctest_html(decl_index: Decl.Index) String {
         return String.init("");
 
     string_result.clearRetainingCapacity();
-    file_source_html(decl.file, &string_result, doctest_ast_node, .{}) catch |err| {
+    file_source_html(decl.file, &string_result, doctest_ast_node, .{
```
Comment (sno2) [inline lib/docs/wasm/main.zig:545]: This adds the same behavior to rendering doctests.

**Comment by nektro** [comment]
imo it should print option 2. just the decl and whitespace trimmed while preserving indentation.

**Comment by sno2** [comment]
> imo it should print option 2. just the decl and whitespace trimmed while preserving indentation.

Hmm, I think we disagree on what would be the actual "Source Code." Although, the current setup doesn't show any of the tokens on the last line after the terminator, so it would match that setup if we did option 2.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks for the patch, @sno2. This discussion has convinced me that simply including the entire line is not good enough, and the original required behavior I specified in #19293 is the required behavior.

#### Case 1

Input:
```zig
    fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 2

Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 3

Input: 
```zig
    fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:

```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

#### Case 4


Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

**Comment by nektro** [comment]
imo the existence of zig fmt's api in the stdlib should mean that with:

```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
return a + b;
    }
```

when viewing the source of `add` is still rendered as:

```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

**Comment by sno2** [comment]
If anyone else would like to work on this- be my guest. I have exams to study for until halfway through next week so I will be busy until then. Will re-open next week.

**Pull request closed without merging**